### PR TITLE
bpo-39199: Add descriptions of non-deprecated AST nodes to the AST module documentation

### PR DESCRIPTION
**Pull request by pablogsal**
These new descriptions are adapted from https://greentreesnakes.readthedocs.io

https://bugs.python.org/issue39199

### CODE REVIEW

**Comment by willingc** [comment]
@pablogsal Just getting back from vacation. Will look at this in the next day or so. Happy New Year too!

File: Doc/library/ast.rst [hunk 1]
```diff
@@ -108,6 +112,1390 @@ Node classes
...
+   * ``value`` is any expression node (such as a literal, a variable, or a
+     function call).
+   * ``conversion`` is an integer:
+
+     * -1: no formatting
+     * 115: ``!s`` string formatting
+     * 114: ``!r`` repr formatting
+     * 97: ``!a`` ascii formatting
+
+   * ``format_spec`` is a :class:`JoinedStr` node reprensenting the formatting
```
Comment (tirkarthi) [inline Doc/library/ast.rst]: Minor typo : 

```suggestion
   * ``format_spec`` is a :class:`JoinedStr` node representing the formatting
```

File: Doc/library/ast.rst [hunk 2]
```diff
@@ -108,6 +112,1390 @@ Node classes
    :class:`ast.NameConstant` and :class:`ast.Ellipsis` are still available,
    but they will be removed in future Python releases.
 
+Literals
+^^^^^^^^
+
+.. class:: Constant(value)
+
+   A constant value. The ``value`` attribute of the objects contains the
```
Comment (willingc) [inline Doc/library/ast.rst]:
```suggestion
   A constant value. The ``value`` attribute of the ``Constant`` literal contains the
```

File: Doc/library/ast.rst [hunk 3]
```diff
@@ -108,6 +112,1390 @@ Node classes
...
+                    elts=[
+                        Constant(value=1, kind=None),
+                        Constant(value=2, kind=None),
+                        Constant(value=3, kind=None)],
+                    ctx=Load()))],
+        type_ignores=[])
+
+.. class:: Set(elts)
+
+   A set. ``elts`` holds a list of nodes representing the elements. ::
```
Comment (willingc) [inline Doc/library/ast.rst]:
```suggestion
   A set. ``elts`` holds a list of nodes representing the set's elements. ::
```

File: Doc/library/ast.rst [hunk 4]
```diff
@@ -108,6 +112,1390 @@ Node classes
...
+   Binary operator tokens.
+
+.. class:: BoolOp(op, values)
+
+   A boolean operation, 'or' or 'and'. ``op`` is :class:`Or` or
+   :class:`And`. ``values`` are the values involved. Consecutive operations
+   with the same operator, such as ``a or b or c``, are collapsed into one node
+   with several values.
+
+   This doesn't include ``not``, which is a :class:`UnaryOp`.
```
Comment (willingc) [inline Doc/library/ast.rst:436]: Nice! Good clarification.

File: Doc/library/ast.rst [hunk 5]
```diff
@@ -108,6 +112,1390 @@ Node classes
...
+.. note::
+   Optional clauses such as ``else`` are stored as an empty list if they're
+   not present.
+
+.. class:: If(test, body, orelse)
+
+   An ``if`` statement. ``test`` holds a single node, such as a :class:`Compare`
+   node. ``body`` and ``orelse`` each hold a list of nodes.
+
+   ``elif`` clauses don't have a special representation in the AST, but rather
```
Comment (willingc) [inline Doc/library/ast.rst:1083]: +1

**Review by willingc — APPROVED** [review]
@pablogsal This is wonderful. Great job. A few little nits but this looks ready to merge. Thanks!

**Comment by JulienPalard** [comment]
LGTM, @pablogsal take a look at @willingc comments and it's good for me.

**Comment by takluyver** [comment]
I was slightly surprised when I consulted the AST module docs today to see that a large part of Green Tree Snakes has been incorporated with no visible attribution (in the docs themselves - I can see it's acknowledged in this PR).

I am pleased to see it there - I think this should have been part of the official AST docs in the first place, and I'm sure it will be kept up to date better here than as a random separate project. But at the same time, a fair amount of effort (mostly other contributors' in the last few years) went into creating and curating those descriptions. I would have expected someone to ask before republishing it.

I never actually got round to adding any kind of license to the GTS repo, though that's due to oversight & laziness rather than any objection to the idea. If I had, it would likely be something like CC-BY or BSD, though I'd be happy to waive the attribution requirements (as much as I can - other contributors retain their rights) if that was even a small obstacle to improving the Python docs.

If you think it's appropriate, I would be delighted to see a little note somewhere by this section mentioning something like "these descriptions of node classes were adapted from [Green Tree Snakes](https://greentreesnakes.readthedocs.io/en/latest/)". I think that's the easiest way to acknowledge something that many people have worked on. If there's no natural way to do this visibly in the documentation, I'd settle for the compromise of putting it in a comment in the rst source.

**Comment by pablogsal** [comment]
Thanks a lot for reaching out @takluyver ! 

> If you think it's appropriate, I would be delighted to see a little note somewhere by this section mentioning something like "these descriptions of node classes were adapted from [Green Tree Snakes](https://greentreesnakes.readthedocs.io/en/latest/)". I think that's the easiest way to acknowledge something that many people have worked on. If there's no natural way to do this visibly in the documentation, I'd settle for the compromise of putting it in a comment in the rst source.

Absolutely, I think it makes sense. Let's add some visible attribution on the documentation itself! Would you mind reviewing the PR once I make it?

**Comment by takluyver** [comment]
Thanks! I'd be happy to review a PR.

**Comment by pablogsal** [comment]
> Thanks! I'd be happy to review a PR.

Opened https://github.com/python/cpython/pull/24727

**Pull request merged**